File: frontend/src/store/reportes.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '../services/api'
import type { ReporteDetail, ReporteCreate, ReporteUpdate, ReporteResponse, PaginatedReportes, Paginacion } from '../types'

export const useReportesStore = defineStore('reportes', () => {
  const reportes = ref<ReporteDetail[]>([])
  const reporteActual = ref<ReporteDetail | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)
  const paginacion = ref<Paginacion>({
    pagina: 1,
    tamanio: 10,
    total_elementos: 0,
    total_paginas: 0,
  })

  const filtros = ref<{
    estado?: string
    tipo?: string
    entidad_tipo?: string
  }>({})

  async function fetchReportes(overrideFiltros?: {
    estado?: string
    tipo?: string
    entidad_tipo?: string
    pagina?: number
  }) {
    loading.value = true
    error.value = null
    try {
      const params: Record<string, string | number | boolean | undefined | null> = {
        estado: overrideFiltros?.estado ?? filtros.value.estado,
        tipo: overrideFiltros?.tipo ?? filtros.value.tipo,
        entidad_tipo: overrideFiltros?.entidad_tipo ?? filtros.value.entidad_tipo,
        pagina: overrideFiltros?.pagina ?? paginacion.value.pagina,
        tamanio: paginacion.value.tamanio,
      }
      const result = await api.get<PaginatedReportes>('/reportes', params)
      reportes.value = result.data
      paginacion.value = result.paginacion
    } catch (e: unknown) {
      error.value = e instanceof Error ? e.message : 'Error al cargar reportes'
    } finally {
      loading.value = false
    }
  }

  async function fetchReporte(id: number) {
    loading.value = true
    error.value = null
    try {
      reporteActual.value = await api.get<ReporteDetail>(`/reportes/${id}`)
    } catch (e: unknown) {
      error.value = e instanceof Error ? e.message : 'Error al cargar reporte'
    } finally {
      loading.value = false
    }
  }

  async function createReporte(data: ReporteCreate) {
    loading.value = true
    error.value = null
    try {
      await api.post<ReporteResponse>('/reportes', data)
    } catch (e: unknown) {
      error.value = e instanceof Error ? e.message : 'Error al enviar reporte'
      throw e
    } finally {
      loading.value = false
    }
  }

  async function updateReporte(id: number, data: ReporteUpdate) {
    loading.value = true
    error.value = null
    try {
      const actualizado = await api.patch<ReporteDetail>(`/reportes/${id}`, data)
      const idx = reportes.value.findIndex((r) => r.id === id)
      if (idx !== -1) {
        reportes.value[idx] = actualizado
      }
      if (reporteActual.value?.id === id) {
        reporteActual.value = actualizado
      }
      return actualizado
    } catch (e: unknown) {
      error.value = e instanceof Error ? e.message : 'Error al actualizar reporte'
      throw e
    } finally {
      loading.value = false
    }
  }

  function setFiltros(nuevosFiltros: { estado?: string; tipo?: string; entidad_tipo?: string }) {
    filtros.value = { ...nuevosFiltros }
    paginacion.value.pagina = 1
    fetchReportes()
  }

  function irPagina(pagina: number) {
    paginacion.value.pagina = pagina
    fetchReportes()
  }

  return {
    reportes,
    reporteActual,
    loading,
    error,
    paginacion,
    filtros,
    fetchReportes,
    fetchReporte,
    createReporte,
    updateReporte,
    setFiltros,
    irPagina,
  }
})

File: backend/app/routers/reportes.py
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_admin_user, get_current_user
from app.models.usuario import Usuario
from app.schemas.comun import ErrorResponse
from app.schemas.reporte import (
    PaginatedReporteResponse,
    ReporteCreate,
    ReporteDetail,
    ReporteResponse,
    ReporteUpdate,
)
from app.schemas.usuario import UsuarioResponse
from app.services.reporte_service import (
    FiltrosReporte,
    create_reporte,
    get_reporte,
    get_reportes,
    update_reporte,
)

router = APIRouter(prefix="/reportes", tags=["Reportes"])


@router.post(
    "",
    response_model=ReporteResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Crear un reporte",
    responses={
        422: {"model": ErrorResponse, "description": "Error de validacion"},
    },
)
async def crear_reporte(
    data: ReporteCreate,
    db: AsyncSession = Depends(get_db),
    usuario: Usuario = Depends(get_current_user),
) -> Any:
    """Crea un nuevo reporte.

    Requiere autenticacion. El reporte se asocia automaticamente al usuario
    que realiza la denuncia.
    """
    reporte = await create_reporte(
        db, data, usuario_id=usuario.id
    )
    return ReporteResponse(
        id=reporte.id,
        tipo=reporte.tipo,
        entidad_tipo=reporte.entidad_tipo,
        entidad_id=reporte.entidad_id,
        descripcion=reporte.descripcion,
        detalle_contacto=reporte.detalle_contacto,
        estado=reporte.estado,
        fecha_creacion=reporte.fecha_creacion,
    )


@router.get(
    "",
    response_model=PaginatedReporteResponse,
    summary="Listar reportes (solo admin)",
    responses={
        401: {"model": ErrorResponse, "description": "No autenticado"},
        403: {"model": ErrorResponse, "description": "Acceso denegado"},
    },
)
async def listar_reportes(
    pagina: int = Query(1, ge=1, description="Numero de pagina"),
    tamanio: int = Query(20, ge=1, le=100, description="Elementos por pagina"),
    estado: Optional[str] = Query(None, description="Filtrar por estado"),
    tipo: Optional[str] = Query(None, description="Filtrar por tipo de reporte"),
    entidad_tipo: Optional[str] = Query(
        None, description="Filtrar por tipo de entidad"
    ),
    db: AsyncSession = Depends(get_db),
    admin: Usuario = Depends(get_admin_user),
) -> Any:
    """Retorna la lista de reportes paginados.

    Solo accesible por usuarios administradores.
    Permite filtrar por estado, tipo y entidad_tipo.
    """
    filters = FiltrosReporte(
        pagina=pagina,
        tamanio=tamanio,
        estado=estado,
        tipo=tipo,
        entidad_tipo=entidad_tipo,
    )
    reportes, total_elementos, pagina_actual, total_paginas = await get_reportes(
        db, filters
    )

    data = [
        ReporteDetail(
            id=r.id,
            tipo=r.tipo,
            entidad_tipo=r.entidad_tipo,
            entidad_id=r.entidad_id,
            descripcion=r.descripcion,
            detalle_contacto=r.detalle_contacto,
            estado=r.estado,
            fecha_creacion=r.fecha_creacion,
            usuario=UsuarioResponse.model_validate(r.usuario) if r.usuario else None,
            comentario_admin=r.comentario_admin,
            resuelto_por=r.resuelto_por,
            fecha_actualizacion=r.fecha_actualizacion,
        )
        for r in reportes
    ]

    return PaginatedReporteResponse(
        data=data,
        paginacion={
            "pagina": pagina_actual,
            "tamanio": tamanio,
            "total_elementos": total_elementos,
            "total_paginas": total_paginas,
        },
    )


@router.get(
    "/{reporte_id}",
    response_model=ReporteDetail,
    summary="Obtener detalle de un reporte (solo admin)",
    responses={
        401: {"model": ErrorResponse, "description": "No autenticado"},
        403: {"model": ErrorResponse, "description": "Acceso denegado"},
        404: {"model": ErrorResponse, "description": "Reporte no encontrado"},
    },
)
async def obtener_reporte(
    reporte_id: int,
    db: AsyncSession = Depends(get_db),
    admin: Usuario = Depends(get_admin_user),
) -> Any:
    """Retorna el detalle completo de un reporte.

    Solo accesible por usuarios administradores.
    """
    reporte = await get_reporte(db, reporte_id)
    if not reporte:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Reporte no encontrado",
        )

    return ReporteDetail(
        id=reporte.id,
        tipo=reporte.tipo,
        entidad_tipo=reporte.entidad_tipo,
        entidad_id=reporte.entidad_id,
        descripcion=reporte.descripcion,
        detalle_contacto=reporte.detalle_contacto,
        estado=reporte.estado,
        fecha_creacion=reporte.fecha_creacion,
        usuario=UsuarioResponse.model_validate(reporte.usuario)
        if reporte.usuario
        else None,
        comentario_admin=reporte.comentario_admin,
        resuelto_por=reporte.resuelto_por,
        fecha_actualizacion=reporte.fecha_actualizacion,
    )


@router.patch(
    "/{reporte_id}",
    response_model=ReporteDetail,
    summary="Actualizar estado y comentario de un reporte (solo admin)",
    responses={
        401: {"model": ErrorResponse, "description": "No autenticado"},
        403: {"model": ErrorResponse, "description": "Acceso denegado"},
        404: {"model": ErrorResponse, "description": "Reporte no encontrado"},
        422: {"model": ErrorResponse, "description": "Error de validacion"},
    },
)
async def actualizar_reporte(
    reporte_id: int,
    data: ReporteUpdate,
    db: AsyncSession = Depends(get_db),
    admin: Usuario = Depends(get_admin_user),
) -> Any:
    """Actualiza el estado y/o comentario de un reporte.

    Solo accesible por usuarios administradores.
    Al marcar como 'resuelto' o 'rechazado' se registra automaticamente
    el administrador que realizo la accion.
    """
    reporte = await update_reporte(db, reporte_id, data, admin_id=admin.id)
    if not reporte:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Reporte no encontrado",
        )

    return ReporteDetail(
        id=reporte.id,
        tipo=reporte.tipo,
        entidad_tipo=reporte.entidad_tipo,
        entidad_id=reporte.entidad_id,
        descripcion=reporte.descripcion,
        detalle_contacto=reporte.detalle_contacto,
        estado=reporte.estado,
        fecha_creacion=reporte.fecha_creacion,
        usuario=UsuarioResponse.model_validate(reporte.usuario)
        if reporte.usuario
        else None,
        comentario_admin=reporte.comentario_admin,
        resuelto_por=reporte.resuelto_por,
        fecha_actualizacion=reporte.fecha_actualizacion,
    )
